util/math_reference.h

```diff
@@ -662,955 +371,1326 @@ T popcount(T x) {
+      [](unsigned short x) { return nan(x); }, a);
+}
+
+// FIXME: AdaptiveCpp does not support marray
+#ifndef SYCL_CTS_COMPILING_WITH_ADAPTIVECPP
+template <size_t N>
+sycl::marray<sycl::half, N> nan(sycl::marray<unsigned short, N> a) {
+  return sycl_cts::math::run_func_on_marray<sycl::half, unsigned short, N>(
+      [](unsigned short x) { return nan(x); }, a);
+}
+#endif
+
+sycl::half nextafter(sycl::half a, sycl::half b);
+sycl::half sinpi(sycl::half a);
+sycl::half tanpi(sycl::half a);
+
+sycl::half fast_dot(float p0);
+sycl::half fast_dot(sycl::float2 p0);
+sycl::half fast_dot(sycl::float3 p0);
+sycl::half fast_dot(sycl::float4 p0);
+
+// FIXME: AdaptiveCpp does not support marray
+#ifndef SYCL_CTS_COMPILING_WITH_ADAPTIVECPP
+sycl::half fast_dot(sycl::mfloat2 p0);
+sycl::half fast_dot(sycl::mfloat3 p0);
+sycl::half fast_dot(sycl::mfloat4 p0);
+#endif
+
+#endif  // SYCL_CTS_ENABLE_HALF_TESTS
+
+#if SYCL_CTS_ENABLE_DOUBLE_TESTS
+
+template <>
+struct higher_accuracy<double> {
+  using type = long double;
+};
+
+double bitselect(double a, double b, double c);
+double degrees(double a);
+sycl_cts::resultRef<double> mix(const double a, const double b, const double c);
+double radians(double a);
+double step(double a, double b);
+sycl_cts::resultRef<double> smoothstep(double a, double b, double c);
+double sign(double a);
+
+double acospi(double a);
+double asinpi(double a);
+double atanpi(double a);
+double atan2pi(double a, double b);
+double cospi(double a);
+double fma(double a, double b, double c);
+double fract(double a, double* b);
+double nan(unsigned long a);
+double nan(unsigned long long a);
+
+template <typename T, int N>
+std::enable_if_t<std::is_same_v<unsigned long, T> ||
+                     std::is_same_v<unsigned long long, T>,
+                 sycl::vec<double, N>>
+nan(sycl::vec<T, N> a) {
+  return sycl_cts::math::run_func_on_vector<double, T, N>(
+      [](T x) { return nan(x); }, a);
+}
+
+// FIXME: AdaptiveCpp does not support marray
+#ifndef SYCL_CTS_COMPILING_WITH_ADAPTIVECPP
+template <typename T, size_t N>
+std::enable_if_t<std::is_same_v<unsigned long, T> ||
+                     std::is_same_v<unsigned long long, T>,
+                 sycl::marray<double, N>>
+nan(sycl::marray<T, N> a) {
+  return sycl_cts::math::run_func_on_marray<double, T, N>(
+      [](T x) { return nan(x); }, a);
+}
+#endif
+
+double sinpi(double a);
+double tanpi(double a);
+
+sycl::double4 cross(sycl::double4 p0, sycl::double4 p1);
+sycl::double3 cross(sycl::double3 p0, sycl::double3 p1);
+
+// FIXME: AdaptiveCpp does not support marray
+#ifndef SYCL_CTS_COMPILING_WITH_ADAPTIVECPP
+sycl::mdouble4 cross(sycl::mdouble4 p0, sycl::mdouble4 p1);
+sycl::mdouble3 cross(sycl::mdouble3 p0, sycl::mdouble3 p1);
+#endif
+
+#endif  // SYCL_CTS_ENABLE_DOUBLE_TESTS
+
+// sycl::vec overloads of the above. These must be defined after any
+// non-templated conditional half / double overloads for correct resolution with
+// DPC++.
```

Comment:
Could you clarify what this comment says, please? Is this something DPC++ compiler requires or C++ language rules?

Reply:
I looked into this and it's not DPC++ specific. For a function to participate in overload resolution it must be declared prior to the call site, not just prior to where the caller is instantiated (see https://godbolt.org/z/WxzzKaasY). Probably because `fn` in my example is not a dependent name?

I've clarified in the comment what the requirements are and that it's not related to DPC++.